rfd/0147-repo-fork-syncing.md

```diff
@@ -0,0 +1,60 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+ 
+# RFD 144 - Automatic Repo Syncing
+ 
+## What
+ 
+This RFD describes how we can sync a repo continuously and in an automatic way.  This can be used to provide private copies of a repo.
+ 
+## Why
+ 
+We currently face a few frictions which could be addressed through a more robust repo sync mechanism:
+1. Our internal `teleport-private` fork used for security development must be manually updated
+2. Some tooling (Dependabot) does not allow scanning for non-default branches, necessitating manual efforts to manage our release branches
+3. Because much of the CI and scanning configuration is committed into the repo, we face noise from duplicate alerts when we use a repo for testing or as an internal fork
+ 
+## Details
+ 
+As such we need a solution that can provide the following:
+* Automatic and regular updates to a repo to ensure it's in sync
+* The ability to apply changes to the upstream repo (for example disable Actions that we don't want running on the repo copy, or enable additional scanning / actions)
+ 
+### Branch Structure
+ 
+* `main` - This will be the repo's default branch.  This branch will contain the upstream history in addition to commits needed for the sync process or other custom changes on the repo copy
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch.  If there is a need to create a PR in this repo the history should be based off this rather than `main`
```

Comment:
> the history should be based off this rather than `main`

Should PRs also target these `sync/upstream` branches, or will they target the `sync/rebase` branches?  Won't history get wonked up when a PR merges (in either direction)?

E.g. say we have 3 security release fixes that need to be merged over the course of a week (during which we also would have some sync updates).  What would the history look like here?

Reply:
> Should PRs also target these sync/upstream branches, or will they target the sync/rebase branches? Won't history get wonked up when a PR merges (in either direction)?

Short answer: it doesn't matter since we don't merge the PR's in the forks.  There is no bi-directional sync here, so PR's need to be re-submitted from the upstream parent repo.  If a PR was merged, it would be overwritten the next time the sync runs.

But making sure the PR is sourced from the `sync/upstream` branch ensures you can simply push the branch to the upstream without modification to submit that PR.

Reply:
There is a new section `teleport-private Developer Experience` which highlights the specific changes.